### PR TITLE
Wip/hughsie/gudev

### PR DESCRIPTION
**Pull request by hughsie**
I was going to use something like this for the nvme plugin too, if it looks okay.

### CODE REVIEW

**Comment by superm1** [comment]
Probably means you can drop instances of
`G_DEFINE_AUTOPTR_CLEANUP_FUNC(GUdevDevice, g_object_unref)`
across the codebase too right?

File: plugins/thunderbolt-power/fu-plugin-thunderbolt-power.c [hunk 1]
```diff
@@ -11,7 +11,6 @@
 #include <gio/gio.h>
 #include <glib.h>
 #include <glib/gstdio.h>
-#include <gudev/gudev.h>
```
Comment (superm1) [inline plugins/thunderbolt-power/fu-plugin-thunderbolt-power.c]: This plugin should be more than just a header change, it should start using the daemon wide client too I think.

File: plugins/thunderbolt/fu-plugin-thunderbolt.c [hunk 2]
```diff
@@ -16,7 +16,6 @@
 
 #include <gio/gio.h>
 #include <glib.h>
-#include <gudev/gudev.h>
```
Comment (superm1) [inline plugins/thunderbolt/fu-plugin-thunderbolt.c]: likewise this one too

**Comment by hughsie** [comment]
For some perspective, using this as the base branch for the NVMe plugin gets me:

     fu-nvme-device.c |   29 +++++-----------
     fu-nvme-device.h |    2 -
     fu-plugin-nvme.c |   99 +++----------------------------------------------------
     meson.build      |    2 -
     4 files changed, 18 insertions(+), 114 deletions(-)

File: plugins/thunderbolt-power/fu-plugin-thunderbolt-power.c (hunk 1 above)
Comment (hughsie) [reply plugins/thunderbolt-power/fu-plugin-thunderbolt-power.c]: I didn't want to port this until the bolt stuff was done.

**Comment by superm1** [comment]
> For some perspective, using this as the base branch for the NVMe plugin gets me:

Wow.  Sounds like the right thing to do :)

File: plugins/thunderbolt-power/fu-plugin-thunderbolt-power.c (hunk 1 above)
Comment (superm1) [reply plugins/thunderbolt-power/fu-plugin-thunderbolt-power.c]: Okay.

File: plugins/thunderbolt/fu-plugin-thunderbolt.c (hunk 2 above)
Comment (hughsie) [reply plugins/thunderbolt/fu-plugin-thunderbolt.c]: This is slightly tricky; this uses the "changed" signal from udev. I don't really want another vfunc for this, and was pondering a `FuUdevDevice::changed` signal that plugins could connect to.

Comment (superm1) [reply plugins/thunderbolt/fu-plugin-thunderbolt.c]: That sounds like a good idea to me.

Comment (hughsie) [reply plugins/thunderbolt/fu-plugin-thunderbolt.c]: I'll give this a go.

Comment (hughsie) [reply plugins/thunderbolt/fu-plugin-thunderbolt.c]: I've added the changed signal, but the thunderbolt plugin is somewhat tricky to port as it's got the device-wait stuff too. I'm now thinking we should just finish off https://github.com/hughsie/fwupd/pull/329 and use that and FuUdevDevice together. In the meantime I'd like to merge this PR and get the nvme stuff merged before the next tarball.

**Comment by hughsie** [comment]
Merged, with a few extra fixes.

**Pull request closed without merging**